src/sentry_os.c

```diff
@@ -195,8 +195,105 @@ sentry__get_os_context(void)
 }
 #elif defined(SENTRY_PLATFORM_UNIX)
 
+#    include <fcntl.h>
 #    include <sys/utsname.h>
 
+#    if defined(SENTRY_PLATFORM_LINUX)
+#        define OS_RELEASE_MAX_LINE_SIZE 128
+#        define OS_RELEASE_MAX_KEY_SIZE 64
+#        define OS_RELEASE_MAX_VALUE_SIZE 128
+
+static int
+parse_os_release_line(const char *line, char *key, char *value)
+{
+    const char *equals = strchr(line, '=');
+    if (equals == NULL)
+        return 1;
+
+    unsigned long key_length = equals - line;
+    strncpy(key, line, key_length);
+    key[key_length] = 0;
+
+    unsigned long value_length = strlen(equals + 1);
+    strncpy(value, equals + 1, value_length);
+    value[value_length] = 0;
+
+    // some values are wrapped in double quotes
+    if (value[0] == '\"') {
+        value[value_length - 1] = 0;
+        memmove(value, value + 1, value_length);
+    }
+
+    return 0;
+}
+
+#        ifndef SENTRY_UNITTEST
+static
+#        endif
+    sentry_value_t
+    get_linux_os_release(const char *os_rel_path)
+{
+    const int fd = open(os_rel_path, O_RDONLY);
+    if (fd == -1) {
+        return sentry_value_new_null();
+    }
+
+    sentry_value_t os_dist = sentry_value_new_object();
+    char buffer[OS_RELEASE_MAX_LINE_SIZE];
+    ssize_t bytes_read;
+    ssize_t buffer_rest = 0;
+    const char *line = buffer;
+    while ((bytes_read = read(
+                fd, buffer + buffer_rest, sizeof(buffer) - buffer_rest - 1))
+        > 0) {
+        buffer[bytes_read] = 0;
+
+        for (char *p = buffer; *p; ++p) {
+            if (*p != '\n')
+                continue;
+
+            char value[OS_RELEASE_MAX_VALUE_SIZE];
+            char key[OS_RELEASE_MAX_KEY_SIZE];
+            *p = '\0';
+
+            if (parse_os_release_line(line, key, value) == 0) {
+                if (strcmp(key, "ID") == 0) {
+                    sentry_value_set_by_key(
+                        os_dist, "name", sentry_value_new_string(value));
+                }
+
+                if (strcmp(key, "VERSION_ID") == 0) {
+                    sentry_value_set_by_key(
+                        os_dist, "version", sentry_value_new_string(value));
+                }
+
+                if (strcmp(key, "PRETTY_NAME") == 0) {
+                    sentry_value_set_by_key(
+                        os_dist, "pretty_name", sentry_value_new_string(value));
+                }
+            }
+            line = p + 1;
+        }
+
+        // Handle any partial line left at the end of the buffer
+        if (line < buffer + bytes_read) {
+            buffer_rest = buffer + bytes_read - line;
+            memmove(buffer, line, buffer_rest);
+            line = buffer;
+        }
+    }
+
+    if (bytes_read == -1) {
```

Comment:
I would love to cover this, and I implemented it locally with a minimal FUSE setup that allows opening a file but returns an error on any read. I just wonder whether this PR is where we'll add it to the test setup.

Reply:
We could probably change the interface to separate `open()` from `read()`, pass an `fd` instead of a path, and modify the `fd` in between. But that would move the goalposts in non-public interfaces. 

In C unit tests, error handling is a significant part of the uncovered paths, and introducing a heavier approach should help cover most of the related test cases in the codebase. So, whatever approach is taken, this will happen in a different PR 😅.